### ISSUE TITLE
Building a C precompiled header fails; treated as C++

### ISSUE THREAD

**Issue by Akaricchi** [emscripten-core/emscripten#10872]
[`emcc -v` command and output.](https://gist.githubusercontent.com/Akaricchi/3f69cce95a589e6c4147f26a6233da34/raw/bb77612ff2428fd7f3c104ae51c9a40c8921b1a6/gistfile1.txt)

My project is pure C, not C++. There is nothing on the command line to suggest C++ mode as far as I can tell, and I'm invoking `emcc`, not `em++`. Yet for some reason clang thinks I want C++. This doesn't happen with regular Clang 10.

**Comment by sbc100** [comment]
emcc, unlike clang or gcc will default to C++ unless you give it an input file the looks like a C file, or if you pass `-x c`.    This is why the problem only shows up when you compile your header files.

See the comment here for the description of the current behaviour:  https://github.com/emscripten-core/emscripten/blob/ceacba112c655c66a356a8066e9343d74568ea4d/emcc.py#L709

I'd like to change this to match the gcc/clang behaviour if we could do it without breaking to many existing users.

In the mean time maybe you can pass `-x c` to solve your immediate issue?

**Comment by Akaricchi** [comment]
> In the mean time maybe you can pass -x c to solve your immediate issue?

The problem with that is that this command is generated by Meson's built-in support for pre-compiled headers. To work around it locally I'd have to introduce a horrible hack into my build definitions (basically duplicate all of pch logic that Meson is supposed to handle for me), which is just not worth the trouble. I could instead add a yet another special case for Emscripten into Meson and send them a patch, but I think ultimately it would be much better for everyone if emcc behavior matches clang as close as possible.

> This means that emcc will act as a C++ linker when no source files are specified.

Is that the only benefit of this behavior? I think it's reasonable to expect people to invoke `em++` when linking C++ objects.

**Comment by sbc100** [comment]
> Is that the only benefit of this behavior? I think it's reasonable to expect people to invoke `em++` when linking C++ objects.

I agree.  I'm just not sure how many people will be effected.

Internally the test suite uses `emcc` all over the place to compile *and* link C++ programs.  I think step one would be making a switch for this behaviour and fixing all our internal usage.

But wait, I don't understand why you can't add `-x c` to your CFLAGS?  I see all kind of other cflags being uses in the command line you posted (e.g. `-s ASYNCIFY`).   Another alternative is to create an alias or wrapper for emcc that inject `-x c`.   I'm just suggesting these as options in case changing the emcc behaviour proves hard to takes a while, you will want some interim solution.

**Comment by sbc100** [comment]
Oh, we could put this behaviour behind `-s STRICT` initially.

**Comment by Akaricchi** [comment]
> But wait, I don't understand why you can't add -x c to your CFLAGS? I see all kind of other cflags being uses in the command line you posted (e.g. -s ASYNCIFY)

I can't add flags *just* for the pch-generating command. I guess I can add it to my cflags globally, if that won't break anything else. But I'm inclined to just not use pch for emscripten builds until this is fixed; I don't really rebuild for emscripten often enough to justify such hacks in the build system.

> Oh, we could put this behaviour behind -s STRICT initially.

I already build with `-s STRICT`, so that would work nicely for me.

**Comment by Beuc** [comment]
This seems to cause [Python](https://www.beuc.net/python-emscripten/python/dir?ci=tip) to compile improperly, due to configure tests now silently failing:

1.39.11:
```
configure:14482: checking for %lld and %llu printf() format support
configure:14511: renpyweb/emsdk/upstream/emscripten/emcc -c -g -O2 -Werror -Wformat  conftest.c >&5
configure:14511: $? = 0
configure:14565: result: yes
```
1.39.13:
```
configure:14482: checking for %lld and %llu printf() format support
configure:14511: renpyweb/emsdk/upstream/emscripten/emcc -c -g -O2 -Werror -Wformat  conftest.c >&5
clang-11: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Werror,-Wdeprecated]
em++: error: '/.../renpyweb/emsdk/upstream/bin/clang++ -target wasm32-unknown-emscripten -D__EMSCRIPTEN_major__=1 -D__EMSCRIPTEN_minor__=39 -D__EMSCRIPTEN_tiny__=13 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration -Xclang -nostdsysteminc -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/include/libcxx -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/lib/libcxxabi/include -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/lib/libunwind/include -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/include/compat -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/include -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/include/libc -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/lib/libc/musl/arch/emscripten -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/local/include -Xclang -isystem/.../renpyweb/emsdk/.emscripten_cache/wasm/include -c -g -O2 -Werror -Wformat -DEMSCRIPTEN -fignore-exceptions conftest.c -Xclang -isystem/.../renpyweb/emsdk/upstream/emscripten/system/include/SDL -c -o conftest.o -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr' failed (1)
configure:14511: $? = 1
configure: failed program was:
| /* confdefs.h */
[...]
| int
| main ()
| {
| 
|       char *buffer;
|       sprintf(buffer, "%lld", (long long)123);
|       sprintf(buffer, "%lld", (long long)-123);
|       sprintf(buffer, "%llu", (unsigned long long)123);
| 
|   ;
|   return 0;
| }
configure:14565: result: cross -- assuming no
```
That doesn't look good :/

I'm not sure I understand what's the recommendation here?

**Comment by Beuc** [comment]
OK it seems this is fixed with #10951 .
